### ISSUE TITLE
Running tests regularly takes > 10 minutes

### ISSUE THREAD

**Issue by phildini** [pybee/batavia#144]
I think we might be at the point where profiling is needed to see where the testing bottleneck is. Normally I would be ok leaving performance optimizations until the project was more complete, but test builds regularly time out on Travis. This slows down the contributor workflow, and generally makes the project move slower.

I have never done serious python profiling, open to suggestions and/or someone else taking this on.

**Comment by phildini** [comment]
So, looking at tests running on my Mac, phantomjs is using 74% of CPU and 11GB of Memory. 😱 

I'm going to dig into how we can reduce that footprint, ideas welcome.

**Comment by phildini** [comment]
So, I am fairly confident in the path I'm on, which tells me that excessive CPU and memory usage from `phantomjs` is a big part of the slowness and crashing on Travis. I followed a thread which told me to put `page.close()` after the JS had stopped running, but that didn't seem to have an impact on the memory usage. The digging continues.

**Comment by phildini** [comment]
tl;dr: Where in the `TranspileTestCase` should we put the phantom operations so we get one phantom per test class, and which is removed (by force if necessary) after each test class runs?

@freakboy3742 I think I'm at the point where some assistance is needed.

My current working theory is that test builds on Travis break because phantomjs takes all the cpu and memory and then hangs. I was able to prove on my local that phantomjs uses up to **16GB** of memory running the test suite.

Based on some threads I can link to, it appears phantom is _really bad_ at cleaning up after itself when it loads pages. Even if you call `page.close()` phantom just shrugs and tries to eat your RAM.

My proposed solution is, for every test class,

1. Make sure we are instantiating a new phantomjs at the beginning
2. Make sure we are terminating phantomjs with prejudice at the end.

Sidenote: when I start a new phantom session before each individual test and kill it after every test is run, the test suite is _even slower_, but the memory and cpu usage doesn't even register in my Activity Monitor. Starting and killing (killing being the thing I think is missing right now) a phantom process with every test class seems like an ideal middle ground.

**Comment by phildini** [comment]
@freakboy3742 I have something that works locally, and never uses more than 100MB of memory: 

https://github.com/phildini/batavia/commit/2db835cd6619b159da42d0e87754e100fa0fb2df

Unfortunately, while it passes locally, it gets an error on travis: https://travis-ci.org/phildini/batavia/jobs/136620026

Specifically, unable to inject batavia.

Still investigating, wanted to update what I had found.

**Comment by freakboy3742** [comment]
If it helps, the "unable to inject" error usually means one of three things:

1. A path problem, so the test runner can't find the batavia script
2. A dependency problem, usually `jsmin` missing from the test environment.
3. A PhantomJS version problem, where the response from PhantomJS isn't matching the expected response.

Not sure which one you would be hitting here, though.

I have another idea for how to attack this problem - compile all the tests in a class into a single script, run that script, output the results to a single text file, and then grep that file for results. I'm going to have a tinker today and see if I can make this work.

**Comment by glasnt** [comment]
This was resolved in #159
